dalm/datasets/reading_comprehension_generation/utils.py

```diff
@@ -0,0 +1,174 @@
+import csv
+import os
+import re
+import tempfile
+from typing import Dict, Iterator, List, Optional, Tuple
+
+import sentencepiece as spm  # type: ignore[import-untyped]
+from transformers import AutoTokenizer, PreTrainedTokenizerBase
+
+
+def input_generator(directory_or_file: str, csv_column: Optional[str] = None) -> Iterator[Tuple[str, str]]:
+    """
+    Generator that yields the contents of the files in the directory or the CSV column
+    """
+
+    if os.path.isfile(directory_or_file) and not csv_column:
+        raise ValueError("a CSV column must be specified if the input is a file")
+
+    if os.path.isdir(directory_or_file):
+        for file in os.listdir(directory_or_file):
+            file_path = os.path.join(directory_or_file, file)
+            if os.path.isfile(file_path):  # Ensures that we are reading files
+                try:
+                    with open(file_path, "r", encoding="utf-8") as file_contents:
+                        contents = file_contents.read()
+                except UnicodeDecodeError:
+                    with open(file_path, "r", encoding="utf-8", errors="replace") as file_contents:
+                        contents = file_contents.read()
+
+                yield file, contents
```

Comment:
There's a very nasty bug here: if you pass a csv file, the csv aware code below kicks in `directory_or_file.endswith(".csv") and csv_column:` and will work as expected an emit rows of the CSV.

However, if you instead pass a directory that contains CSV files, it will treat the CSV file as raw text and return the _entire text_ (or an entire chunk) rather than rows from the CSV file.

The fix is to treat CSV files the same whether a single CSV is passed in or a directory containing CSV files is passed in.